File: packages/agent-cache/src/adapters/llamaindex.ts
import type { ChatMessage } from "@llamaindex/core/llms";
import type { ContentBlock, LlmCacheParams, TextBlock, BinaryBlock, ToolCallBlock } from "../types";
import type { BinaryNormalizer, BinaryRef } from "../normalizer";
import { defaultNormalizer } from "../normalizer";

/**
 * Minimal interface matching the serializable portion of LlamaIndex's BaseTool.metadata.
 * Only `name` is required; `description` and `parameters` are optional.
 */
export interface ToolMetadataLike {
  name: string;
  description?: string;
  parameters?: Record<string, unknown>;
}

/**
 * Minimal interface for a LlamaIndex tool. Only `metadata` is used for cache keying.
 */
export interface BaseToolLike {
  metadata: ToolMetadataLike;
  [key: string]: unknown;
}

export interface LlamaIndexPrepareOptions {
  model: string;
  normalizer?: BinaryNormalizer;
  temperature?: number;
  topP?: number;
  maxTokens?: number;
  /**
   * Tool definitions to include in the cache key. Pass the same tools array
   * you provide to the LLM call. Only `tool.metadata` (name, description,
   * parameters) is used; the `call` closure is never serialized.
   *
   * Omitting this field falls back to messages-only keying (prior behavior).
   */
  tools?: BaseToolLike[];
}

type AnyDetail = { type: string; text?: string; image_url?: { url: string }; data?: string; mimeType?: string };

function parseInput(input: unknown): unknown {
  if (typeof input === "string") {
    try {
      return JSON.parse(input);
    } catch {
      return { __raw: input };
    }
  }
  return input;
}

async function normalizeDetail(
  part: AnyDetail,
  normalizer: BinaryNormalizer,
): Promise<ContentBlock | null> {
  if (part.type === "text") {
    return { type: "text", text: part.text ?? "" } as TextBlock;
  }

  if (part.type === "image_url" && part.image_url) {
    const url = part.image_url.url;
    let source: BinaryRef["source"];
    let mediaType = "image/*";
    if (url.startsWith("data:")) {
      const semi = url.indexOf(";");
      if (semi > 5) mediaType = url.slice(5, semi);
      source = { type: "base64", data: url };
    } else {
      source = { type: "url", url };
    }
    const ref = await normalizer({ kind: "image", source });
    return { type: "binary", kind: "image", mediaType, ref } as BinaryBlock;
  }

  if (part.type === "file" && part.data) {
    const ref = await normalizer({ kind: "document", source: { type: "base64", data: part.data } });
    return { type: "binary", kind: "document", mediaType: part.mimeType ?? "application/octet-stream", ref } as BinaryBlock;
  }

  if ((part.type === "audio" || part.type === "image") && part.data) {
    const kind = part.type === "audio" ? "audio" as const : "image" as const;
    const ref = await normalizer({ kind, source: { type: "base64", data: part.data } });
    return { type: "binary", kind, mediaType: part.mimeType ?? (kind === "audio" ? "audio/*" : "image/*"), ref } as BinaryBlock;
  }

  return null;
}

export async function prepareParams(
  messages: ChatMessage[],
  opts: LlamaIndexPrepareOptions,
): Promise<LlmCacheParams> {
  const normalizer = opts.normalizer ?? defaultNormalizer;
  const out: Array<{ role: string; content: unknown; toolCallId?: string }> = [];

  for (const msg of messages) {
    const options = (msg as { options?: { toolCall?: unknown[]; toolResult?: { id: string; result: string; isError?: boolean } } }).options;

    // Tool result takes highest priority
    if (options?.toolResult) {
      const tr = options.toolResult;
      out.push({
        role: "tool",
        toolCallId: tr.id,
        content: [{ type: "text", text: tr.result } as TextBlock],
      });
      continue;
    }

    // Map role
    const rawRole = msg.role as string;
    const role = rawRole === "memory" || rawRole === "developer" ? "system" : rawRole;

    // Normalize content
    const blocks: ContentBlock[] = [];
    if (typeof msg.content === "string") {
      if (msg.content !== "") {
        blocks.push({ type: "text", text: msg.content } as TextBlock);
      }
    } else if (Array.isArray(msg.content)) {
      for (const part of msg.content as AnyDetail[]) {
        const b = await normalizeDetail(part, normalizer);
        if (b) blocks.push(b);
      }
    }

    // Append tool calls from options
    if (options?.toolCall) {
      const calls = Array.isArray(options.toolCall) ? options.toolCall : [options.toolCall];
      for (const tc of calls as Array<{ id: string; name: string; input: unknown }>) {
        blocks.push({
          type: "tool_call",
          id: tc.id,
          name: tc.name,
          args: parseInput(tc.input),
        } as ToolCallBlock);
      }
    }

    out.push({ role, content: blocks });
  }

  const result: LlmCacheParams = { model: opts.model, messages: out };
  if (opts.temperature != null) result.temperature = opts.temperature;
  if (opts.topP != null) result.top_p = opts.topP;
  if (opts.maxTokens != null) result.max_tokens = opts.maxTokens;
  if (opts.tools != null && opts.tools.length > 0) {
    result.tools = opts.tools.map((t) => ({
      type: "function",
      function: {
        name: t.metadata.name,
        ...(t.metadata.description != null ? { description: t.metadata.description } : {}),
        ...(t.metadata.parameters != null ? { parameters: t.metadata.parameters } : {}),
      },
    }));
  }

  return result;
}

File: packages/agent-cache-py/betterdb_agent_cache/adapters/llamaindex.py
"""LlamaIndex adapter.

Converts a list of ``ChatMessage`` dicts (as used by LlamaIndex) into
``LlmCacheParams``.

Usage::

    from betterdb_agent_cache.adapters.llamaindex import prepare_params

    params = await prepare_params(messages, model="gpt-4o")
    result = await cache.llm.check(params)
"""
from __future__ import annotations

import json
from dataclasses import dataclass, field
from typing import Any

from ..normalizer import BinaryNormalizer, default_normalizer
from ..types import BinaryBlock, ContentBlock, LlmCacheParams, TextBlock, ToolCallBlock


@dataclass
class LlamaIndexPrepareOptions:
    model: str = ""
    normalizer: BinaryNormalizer = field(default_factory=lambda: default_normalizer)
    temperature: float | None = None
    top_p: float | None = None
    max_tokens: int | None = None
    tools: list[Any] | None = None
    """Tool definitions to include in the cache key.

    Pass the same tools list you provide to the LLM call. Each tool must
    expose a ``metadata`` attribute (or dict key) with at least ``name``,
    and optionally ``description`` and ``parameters``. Only metadata is
    serialized; callable closures are never included.

    Omitting this field falls back to messages-only keying (prior behavior).
    """


def _parse_input(value: Any) -> Any:
    if isinstance(value, str):
        try:
            return json.loads(value)
        except (json.JSONDecodeError, ValueError):
            return {"__raw": value}
    return value


async def _normalize_detail(
    part: dict[str, Any],
    normalizer: BinaryNormalizer,
) -> ContentBlock | None:
    t = part.get("type")

    if t == "text":
        return {"type": "text", "text": part.get("text") or ""}

    if t == "image_url" and part.get("image_url"):
        url: str = part["image_url"]["url"]
        media_type = "image/*"
        if url.startswith("data:"):
            semi = url.find(";")
            if semi > 5:
                media_type = url[5:semi]
            source: dict[str, Any] = {"type": "base64", "data": url}
        else:
            source = {"type": "url", "url": url}
        ref = await normalizer({"kind": "image", "source": source})
        return {"type": "binary", "kind": "image", "mediaType": media_type, "ref": ref}

    if t == "file" and part.get("data"):
        ref = await normalizer({
            "kind": "document",
            "source": {"type": "base64", "data": part["data"]},
        })
        return {
            "type": "binary", "kind": "document",
            "mediaType": part.get("mime_type") or "application/octet-stream",
            "ref": ref,
        }

    if t in ("audio", "image") and part.get("data"):
        kind = "audio" if t == "audio" else "image"
        ref = await normalizer({
            "kind": kind,
            "source": {"type": "base64", "data": part["data"]},
        })
        default_media = "audio/*" if kind == "audio" else "image/*"
        return {
            "type": "binary", "kind": kind,
            "mediaType": part.get("mime_type") or default_media,
            "ref": ref,
        }

    return None


def _extract_tool_metadata(tool: Any) -> dict[str, Any]:
    """Extract serializable metadata from a LlamaIndex BaseTool."""
    if hasattr(tool, "metadata"):
        meta = tool.metadata
    elif isinstance(tool, dict) and "metadata" in tool:
        meta = tool["metadata"]
    else:
        meta = tool  # Already a metadata-like dict

    if hasattr(meta, "name"):
        name = meta.name
        description = getattr(meta, "description", None)
        parameters = getattr(meta, "parameters", None)
    else:
        name = meta.get("name", "")
        description = meta.get("description")
        parameters = meta.get("parameters")

    fn: dict[str, Any] = {"name": name}
    if description is not None:
        fn["description"] = description
    if parameters is not None:
        fn["parameters"] = parameters
    return {"type": "function", "function": fn}


async def prepare_params(
    messages: list[dict[str, Any]],
    opts: LlamaIndexPrepareOptions | None = None,
    *,
    model: str | None = None,
    normalizer: BinaryNormalizer | None = None,
    temperature: float | None = None,
    top_p: float | None = None,
    max_tokens: int | None = None,
    tools: list[Any] | None = None,
) -> LlmCacheParams:
    """Normalise a LlamaIndex message list to ``LlmCacheParams``.

    Either pass an ``LlamaIndexPrepareOptions`` instance or use the keyword
    arguments directly::

        params = await prepare_params(msgs, model="gpt-4o", temperature=0.7)

    To include tool definitions in the cache key (recommended when using tools)::

        params = await prepare_params(msgs, model="gpt-4o", tools=my_tools)
    """
    if opts is None:
        opts = LlamaIndexPrepareOptions(
            model=model or "",
            normalizer=normalizer or default_normalizer,
            temperature=temperature,
            top_p=top_p,
            max_tokens=max_tokens,
            tools=tools,
        )

    norm = opts.normalizer
    out: list[Any] = []

    for msg in messages:
        options: dict[str, Any] = msg.get("options") or {}

        if options.get("tool_result"):
            tr = options["tool_result"]
            out.append({
                "role": "tool",
                "toolCallId": tr["id"],
                "content": [{"type": "text", "text": tr.get("result", "")}],
            })
            continue

        raw_role: str = msg.get("role", "user")
        role = "system" if raw_role in ("memory", "developer") else raw_role

        blocks: list[ContentBlock] = []
        content = msg.get("content")
        if isinstance(content, str):
            if content:
                blocks.append({"type": "text", "text": content})
        elif isinstance(content, list):
            for part in content:
                b = await _normalize_detail(part, norm)
                if b is not None:
                    blocks.append(b)

        tool_calls = options.get("tool_call")
        if tool_calls:
            calls = tool_calls if isinstance(tool_calls, list) else [tool_calls]
            for tc in calls:
                blocks.append({
                    "type": "tool_call",
                    "id": tc["id"],
                    "name": tc["name"],
                    "args": _parse_input(tc.get("input")),
                })

        out.append({"role": role, "content": blocks})

    result: LlmCacheParams = {"model": opts.model, "messages": out}
    if opts.temperature is not None:
        result["temperature"] = opts.temperature
    if opts.top_p is not None:
        result["top_p"] = opts.top_p
    if opts.max_tokens is not None:
        result["max_tokens"] = opts.max_tokens
    if opts.tools is not None and len(opts.tools) > 0:
        result["tools"] = [_extract_tool_metadata(t) for t in opts.tools]

    return result
